pkg/activitypub/store/ariesstore/ariesstore.go

```diff
@@ -0,0 +1,454 @@
+	"time"
+
+	ariesstorage "github.com/hyperledger/aries-framework-go/spi/storage"
+	"github.com/trustbloc/edge-core/pkg/log"
+
+	"github.com/trustbloc/orb/pkg/activitypub/store/memstore"
+	"github.com/trustbloc/orb/pkg/activitypub/store/spi"
+	"github.com/trustbloc/orb/pkg/activitypub/store/storeutil"
+	"github.com/trustbloc/orb/pkg/activitypub/vocab"
+)
+
+const (
+	activityTag      = "Activity"
+	objectIRITagName = "ObjectIRI"
+	timeAddedTagName = "TimeAdded"
+)
+
+var logger = log.New("activitypub_store")
+
+// Provider implements an ActivityPub store backed by an Aries storage provider.
+type Provider struct {
+	serviceName     string
+	activityStore   ariesstorage.Store
+	referenceStores map[spi.ReferenceType]ariesstorage.Store
+	actorStore      ariesstorage.Store
+}
+
+// New returns a new ActivityPub storage provider.
+func New(provider ariesstorage.Provider, serviceName string) (*Provider, error) {
+	stores, err := openStores(provider)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open stores: %w", err)
+	}
+
+	return &Provider{
+		serviceName:     serviceName,
+		activityStore:   stores.activities,
+		referenceStores: stores.reference,
+		actorStore:      stores.actor,
+	}, nil
+}
+
+// PutActor stores the given actor.
+func (s *Provider) PutActor(actor *vocab.ActorType) error {
+	logger.Debugf("[%s] Storing actor [%s]", s.serviceName, actor.ID())
+
+	actorBytes, err := json.Marshal(actor)
+	if err != nil {
+		return fmt.Errorf("failed to marshal actor: %w", err)
+	}
+
+	err = s.actorStore.Put(actor.ID().String(), actorBytes)
+	if err != nil {
+		return fmt.Errorf("failed to store actor: %w", err)
+	}
+
+	return nil
+}
+
+// GetActor returns the actor for the given IRI. Returns an ErrNoFound error if the actor is not in the store.
+func (s *Provider) GetActor(iri *url.URL) (*vocab.ActorType, error) { //nolint: dupl // false positive
+	logger.Debugf("[%s] Retrieving actor [%s]", s.serviceName, iri)
+
+	actorBytes, err := s.actorStore.Get(iri.String())
+	if err != nil {
+		if errors.Is(err, ariesstorage.ErrDataNotFound) {
+			return nil, spi.ErrNotFound
+		}
+
+		return nil, fmt.Errorf("unexpected failure while getting actor from store: %w", err)
+	}
+
+	var actor vocab.ActorType
+
+	err = json.Unmarshal(actorBytes, &actor)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal actor bytes: %w", err)
+	}
+
+	return &actor, nil
+}
+
+// AddActivity adds the given activity to the activity store.
+func (s *Provider) AddActivity(activity *vocab.ActivityType) error {
+	logger.Debugf("[%s] Storing activity - Type: %s, ID: %s",
+		s.serviceName, activity.Type(), activity.ID())
+
+	activityBytes, err := json.Marshal(activity)
+	if err != nil {
+		return fmt.Errorf("failed to marshal activity: %w", err)
+	}
+
+	err = s.activityStore.Put(activity.ID().String(), activityBytes,
+		ariesstorage.Tag{
+			Name: activityTag,
+		}, ariesstorage.Tag{
+			Name:  timeAddedTagName,
+			Value: strconv.FormatInt(time.Now().UnixNano(), 10),
```

Comment:
Note here that I used `UnixNano` instead of `Unix`. My main motivation behind this was to ensure that, in the unit tests, two activities added in succession would have different timestamps. With the second-precision of the regular `Unix` timestamp, most of the time two `AddActivity` calls in a row will result in identical timestamps. (alternatively, I could add a time.Sleep(time.Second) in between calls to ensure they get different time stamps... let me know if you have any concerns about the `nano` approach above. CouchDB seems to be able to handle sorting on the larger numbers fine)

Reply:
I guess it's ok. This should work until the year 2262 :)